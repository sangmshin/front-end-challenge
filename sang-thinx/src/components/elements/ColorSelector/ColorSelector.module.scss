@import 'variables';

.root {
  display: flex;
  height: 30px;
  border: 1px solid black;
  margin-top: 20px;

  & > * {
    flex: 1;
    padding: 0 15px;
    font-size: 0.875rem;
  }

  .color-text {
    display: flex;
    align-items: center;
  }

  .color-button {
    display: flex;
    align-items: center;

    .color-name {
      padding-top: 3px;
    }

    .check-mark {
      margin: 5px 0 0 5px;

      svg {
        width: 15px;
      }
    }
  }
}

@media (min-width: $screen-md) {
  .root {

    & > * {
      font-size: 1rem;
    }

    .color-button {

      .color-name {
        padding-top: 0;
      }

      .check-mark {
        margin-top: 2px;
      }
    }
  }
}