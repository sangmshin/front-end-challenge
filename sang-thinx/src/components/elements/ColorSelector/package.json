{
  "name": "ColorSelector",
  "main": "./ColorSelector.js"
}
