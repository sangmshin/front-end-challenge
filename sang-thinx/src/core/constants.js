export const COLOR_HEXCODE_MAP = {
  beige: '#f2d9a5',
  black: '#000',
}