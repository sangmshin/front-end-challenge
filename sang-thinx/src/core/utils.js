const SMALL = 768; /* Small screen / tablet */
const MEDIUM = 1024; /* Medium screen / desktop */
const LARGE = 1200; /* Large screen / wide desktop */
let BREAKPOINT;

/**
 * Finds the current matched media query breakpoint
 *
 * @returns {String} matched media query breakpoint [large, medium, small, xsmall]
 */
export function getCurrentBreakPoint() {
  if (!BREAKPOINT) {
    BREAKPOINT = {
      large: window.matchMedia(`(min-width: ${LARGE}px)`),
      medium: window.matchMedia(`(min-width: ${MEDIUM}px)`),
      small: window.matchMedia(`(min-width: ${SMALL}px)`),
      xsmall: window.matchMedia(`(max-width: ${SMALL - 1}px)`)
    };
  }

  return Object.keys(BREAKPOINT).find(breakpoint => BREAKPOINT[breakpoint].matches);
}

/**
 * Determines if the data is array and has at least one or more items
 * @return {boolean}
 */
export const isArrayWithLength = arr => Array.isArray(arr) && arr.length > 0;

/**
 * A function with no operation. (for testing)
 */
export const noop = () => null;
