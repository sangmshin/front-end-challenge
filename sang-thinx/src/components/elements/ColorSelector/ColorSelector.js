import React, { createRef, useEffect, useState } from 'react';
import { array } from 'prop-types';
import { Button } from 'components/elements';
import { COLOR_HEXCODE_MAP } from 'core/constants';
import styles from './ColorSelector.module.scss';

const ColorSelector = ({ colors }) => {
  const [selectedColor, selectColor] = useState('');
  const [elementRefs, setElementRefs] = useState([]);

  const handleClick = (e, index) => (
    selectColor(elementRefs[index].current.innerText)
  )

  useEffect(() => {
    selectColor(colors[0]);
    setElementRefs(elementRefs => (
      Array(colors.length)
        .fill()
        .map((_, index) => elementRefs[index] || createRef())
    ));
  }, [colors])
    
  const colorOptions = colors.map((color, index) => {
    const isSelected = selectedColor === color;
    const textColor = color === 'black' ? 'white' : 'black';
    const checkMark = (
      <span className={styles['check-mark']}>
        <svg
          style={{fill: textColor}}
          xmlns="http://www.w3.org/2000/svg"
          width="24"
          height="24"
          viewBox="0 0 24 24"
        >
          <path d="M20.285 2l-11.285 11.567-5.286-5.011-3.714 3.716 9 8.728 15-15.285z"/>
        </svg>
      </span>
    );

    return (
      <Button
        ref={elementRefs[index]}
        ariaLabel={`${color} color`}
        className={styles['color-button']}
        key={color}
        onClick={e => handleClick(e, index)}
        role='radio'
        style={{
          backgroundColor: COLOR_HEXCODE_MAP[color],
          color: textColor
        }}
      >
        {color}
        {isSelected && checkMark}
      </Button>
    )
  });

  return (
    <div className={styles.root}>
      <span>color</span>
      {colorOptions}
    </div>
  )
}

ColorSelector.propTypes = {
  colors: array.isRequired
}

export default ColorSelector;

