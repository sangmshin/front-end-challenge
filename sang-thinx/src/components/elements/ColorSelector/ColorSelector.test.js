import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { ColorSelector } from './ColorSelector';
import { COLOR_HEXCODE_MAP } from 'core/constants';

const availableColors = Object.keys(COLOR_HEXCODE_MAP);

describe('Component :: ColorSelector', () => {
  const component = shallow(
    <ColorSelector colors={availableColors} />
  );

  describe('#render', () => {
    it('should match snapshot', () => {
      expect(toJson(component)).toMatchSnapshot();
    });
  });
});
