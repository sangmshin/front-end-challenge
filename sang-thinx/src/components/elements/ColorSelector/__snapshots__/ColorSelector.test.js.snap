// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component :: ColorSelector #render should match snapshot 1`] = `
<div
  className="root"
>
  <span
    className="color-text"
  >
    color
  </span>
  <ForwardRef
    accessibleText=""
    ariaLabel="beige color"
    className="color-button"
    disabled={false}
    key="beige"
    onClick={[Function]}
    role="radio"
    style={
      Object {
        "backgroundColor": "#f2d9a5",
        "color": "black",
      }
    }
    type="button"
  >
    <span
      className="color-name"
    >
      beige
    </span>
  </ForwardRef>
  <ForwardRef
    accessibleText=""
    ariaLabel="black color"
    className="color-button"
    disabled={false}
    key="black"
    onClick={[Function]}
    role="radio"
    style={
      Object {
        "backgroundColor": "#000",
        "color": "white",
      }
    }
    type="button"
  >
    <span
      className="color-name"
    >
      black
    </span>
  </ForwardRef>
</div>
`;
